core/shared/src/main/scala/scalaz/zio/Managed.scala

```diff
@@ -11,44 +11,35 @@ package scalaz.zio
  * has been consumed, the resource will not be valid anymore and may fail with
  * some checked error, as per the type of the functions provided by the resource.
  */
-sealed abstract class Managed[+E, +R] extends Serializable { self =>
-  type R0 <: R
+final case class Managed[+E, +R](setup: IO[E, Managed.UseToken[E, R]]) { self =>
+  import Managed.UseToken
 
-  protected def acquire: IO[E, R0]
-
-  protected def release: R0 => IO[Nothing, _]
-
-  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A]
+  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A] =
+    setup.flatMap { token =>
+      token.acquire.flatMap(f).ensuring(token.release)
+    }
 
   final def use_[E1 >: E, A](f: IO[E1, A]): IO[E1, A] =
     use(_ => f)
 
   final def map[R1](f0: R => R1): Managed[E, R1] =
-    new Managed[E, R1] {
-      type R0 = R1
-
-      protected def acquire: IO[E, R1] = IO.never
-
-      protected def release: R1 => IO[Nothing, Unit] = _ => IO.unit
-
-      final def use[E1 >: E, A](f: R1 => IO[E1, A]): IO[E1, A] =
-        self.use(r => f(f0(r)))
+    Managed[E, R1] {
+      self.setup.map(token => token.copy(acquire = token.acquire.map(f0)))
     }
 
   final def flatMap[E1 >: E, R1](f0: R => Managed[E1, R1]): Managed[E1, R1] =
-    new Managed[E1, R1] {
-      type R0 = R1
-
-      protected def acquire: IO[E1, R1] = IO.never
-
-      protected def release: R1 => IO[Nothing, Unit] = _ => IO.unit
-
-      final def use[E2 >: E1, A](f: R1 => IO[E2, A]): IO[E2, A] =
-        self.use { r =>
-          f0(r).use { r1 =>
-            f(r1)
-          }
-        }
+    Managed[E1, R1] {
+      Ref.make[List[IO[Nothing, Any]]](List()).map { finalizers =>
+        UseToken(
+          acquire = for {
+            tokenR  <- self.setup
+            r       <- tokenR.acquire.ensuring(finalizers.update(fs => tokenR.release :: fs))
```

Comment:
Hmm, I modeled this after studying how `bracket` works, but I did not understand that `setup` should be the critical point.

Is something like this what you had in mind?
```scala
  final def flatMap[E1 >: E, R1](f0: R => Managed[E1, R1]): Managed[E1, R1] =
    Managed[E1, R1] {
      Ref.make[List[IO[Nothing, Any]]](List()).map { finalizers =>
        UseToken(
          acquire = for {
            rRelease <- Ref.make[Option[IO[Nothing, Any]]](None)
            tokenR  <- self.setup
                           .flatMap(token => rRelease.set(Some(token.release)).const(token))
                           .ensuring {
                             for {
                               release <- rRelease.get
                               _ <- finalizers.update(fs => release ++ fs)
                             } yield ()
                           }
            r       <- tokenR.acquire
            r1Release <- Ref.make[Option[IO[Nothing, Any]]](None)
            tokenR1 <- f0(r).setup
                            .flatMap(token => r1Release.set(Some(token.release)).const(token))
                            .ensuring {
                              for {
                                release <- r1Release.get
                                _ <- finalizers.update(fs => release ++ fs)
                              } yield ()
                            }
            r1      <- tokenR1.acquire
          } yield r1,
          release = finalizers.get.flatMap(IO.collectAll).uninterruptible
        )
      }
    }
```

My problem here is that we could still get interrupted between running `setup` and the release updating. However, we don't want to run `self.setup.flatMap(...)` uninterruptibly, because whoever constructs the `Managed` should decide if setup should be interruptible. And finally we have to flatMap after `setup` to get the token, because `ensuring` doesn't give us the result of the IO.

Reply:
Also it'd be good to know how would you construct the `Managed` for `Semaphore#acquireN` so I could get a sense of what goes into the `setup`/`acquire`/`release` IOs.

Reply:
It's OK for `setup` to be uninterruptible. It's the actual "acquire" that should be interruptible. According to contract, `setup` should be fast and ideally synchronous; for resources like files, it won't do anything, except `IO.succeed` the normal acquire / release handlers. For concurrent data structures, it will allocate a promise (probably) and stuff that into a structure; which is quite fast. Then the acquire could take some time and that will be interruptible for all concurrent data structures, and not for externally-effectful resources.

Reply:
Thanks! That was the missing piece :-)

Reply:
It's hard to comment on code here but let me try:

```scala
final def flatMap[E1 >: E, R1](f0: R => Managed[E1, R1]): Managed[E1, R1] =
    Managed[E1, R1] {
      // There's no need for a List, an IO[Nothing, Any] is fine, you can use *> to chain them
      Ref.make[List[IO[Nothing, Any]]](List()).map { finalizers =>
        UseToken(
          acquire = for {
            // A simpler pattern has emerged: just store IO.unit to get rid of Option
            rRelease <- Ref.make[Option[IO[Nothing, Any]]](None)
            // The setup and storage of the release action should be done uninterruptibly
            tokenR  <- self.setup
                           .flatMap(token => rRelease.set(Some(token.release)).const(token))
                           .ensuring {
                             for {
                               // Why do you have rRelease and r1Release?
                               // I think you just need `finalizers`, you can append to that
                               release <- rRelease.get
                               _ <- finalizers.update(fs => release ++ fs)
                             } yield ()
                           }
            r       <- tokenR.acquire
            r1Release <- Ref.make[Option[IO[Nothing, Any]]](None)
            tokenR1 <- f0®.setup
                            // Ditto for all the above comments
                            .flatMap(token => r1Release.set(Some(token.release)).const(token))
                            .ensuring {
                              for {
                                release <- r1Release.get
                                _ <- finalizers.update(fs => release ++ fs)
                              } yield ()
                            }
            r1      <- tokenR1.acquire
          } yield r1,
          release = finalizers.get.flatMap(IO.collectAll).uninterruptible
        )
      }
    }
```

Reply:
Great feedback, thanks. I'll fix that up.

Reply:
@iravid Did you see `prepare`? It's [here](https://github.com/scalaz/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/Semaphore.scala#L35).

It currently returns an `Acquisition` but we'll refactor that to return a `UseToken`, but it's the same either way: a pair, consisting of an acquire action and a release action. The generation of the acquisition token cannot fail and happens right away, it's ok for that to be uninterruptible; it's the waiting on the acquisition via `acquire` that needs to be interruptible.

I still need to look at how you emulated bracket in terms of this formulation. I'll do that after the flatMap is done & changes propagated.

Thanks so much for working on this. It's not an easy ticket, but on the bright side, you'll become a master of interruption. :)

Reply:
Yeah, I did see it, the fact that it was used in a bracket in `withPermit` threw me off because I thought it was meant to be interruptible. But it makes sense now.

Reply:
Yes, the `setup` (or preparation, in its terminology) is not interruptible (that's why it uses bracket), because that would be bad, but the actual waiting around for the permits becomes interruptible.

In a way, you can think of this version of `Managed` as "squeezing" out the allocation into a hopefully tinier non-interruptible segment, followed by a possibly indefinite (async) acquisition segment; in an effort to make the whole thing more interruptible (because before, acquisition was purely uninterruptible and not broken down into two segments).